common/lib/xmodule/xmodule/js/src/problem/edit.coffee

```diff
@@ -574,8 +580,75 @@ class @MarkdownEditingDescriptor extends XModule.Descriptor
       }
 
       // make all elements descendants of a single problem element
-      xml = '<problem>\n' + xml + demandhints + '\n</problem>';
+      xml = xml + demandhints;
+
+      debugger
+      // make selector to search responsetypes in xml
+      var responseTypesSelector = responseTypes.join(", ");
+
+      // these will be placed at outside the end of responsetype
+      var independentTagNames = ['solution', 'demandhint'];
+      var independentTagNodes = [];
+
+      // make temporary xml
+      var $xml = $($.parseXML('<prob>' + xml + '</prob>'));
+      responseType = $xml.find(responseTypesSelector);
+
+      // convert if there is only one responsetype
+      if (responseType.length === 1) {
+        var inputtype = responseType[0].firstElementChild
+        var before = true;
+
+        _.each($xml.find('prob').children(), function(child, index){
+            // we don't want to add the responsetype again into new xml
+            if (responseType[0].nodeName === child.nodeName) {
+                before = false;
+                return;
+            }
+
+            // replace <p> tag for question title with <label> tag
+            if (child.hasAttribute('class') && child.getAttribute('class') === 'qtitle') {
+                var label = document.createElement("label");
+                label.appendChild(document.createTextNode(child.textContent));
+                child = label;
+            }
+
+            if (_.contains(independentTagNames, child.nodeName)) {
+                independentTagNodes.push(child)
+                return;
+            }
+
+            if (before) {
+                responseType[0].insertBefore(child, inputtype);
+            } else {
+                responseType[0].appendChild(child);
+            }
+        })
+        var serializer = new XMLSerializer();
+
+        // combine responsetype and independent tags
+        xml = serializer.serializeToString(responseType[0]) + '\n\n' + _.map(independentTagNodes, function(node){
+            return serializer.serializeToString(node)
+        }).join('\n\n');
+
+        // remove xmlns attribute added by the serializer
+        xml = xml.replace(/xmlns=['"].*?['"]/gi, "");
+
+        // TODO! Fix xml indentation -- XMLSerializer messes the indentation of XML
+        // add newline at end of each ending tag to make the xml looks better
+        xml = xml.replace(/(\<\/.*?\>)(\<.*?\>)/gi, "$1\n$2");
+      }
 
+      // remove class attribute added on <p> tag for question title
+      xml = xml.replace(/class=\'qtitle\'/gi, "");
       return xml;
```

Comment:
can we move this logic to another function as these changes are part of one concrete thing. This function is already too long, moving this logic into new function will improve readability.